orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py

```diff
@@ -454,10 +340,29 @@ def testToyBertCheckpointLoadZero():
     actual_eval_loss = trainer.eval_step(input_ids, segment_ids, input_mask, masked_lm_labels, next_sentence_labels)
     actual_eval_loss = actual_eval_loss.cpu().numpy().item(0)
 
+    checkpoint_files = sorted(checkpoint._list_checkpoint_files(ckpt_dir, ckpt_prefix))
+    loaded_state_dict = checkpoint.experimental_state_dict(trainer)
+    loaded_state_dict = checkpoint._split_state_dict(loaded_state_dict)
+
+    for f in checkpoint_files:
+        rank_state_dict = torch.load(f, map_location=torch.device("cpu"))
+        rank_state_dict = checkpoint._split_state_dict(rank_state_dict['model'])
+
+        for k,v in rank_state_dict['fp16_param'].items():
+            fp32_name = k.split('_fp16')[0]
+            assert_allclose(v, loaded_state_dict['fp32_param'][fp32_name], rtol=1e-02, atol=1e-02)
+            if mixedprecision_enabled:
+                assert_allclose(v, loaded_state_dict['fp16_param'][k])
+        
+        for k,v in rank_state_dict['optimizer'].items():
+            if k in loaded_state_dict['optimizer']:
+                assert_allclose(v, loaded_state_dict['optimizer'][k])
```

Comment:
this assert tests the loading, but it does not test the aggregation. May we have unit test that gives dummy input to the aggregation function, and checks the aggregated output against what we expect.
For example here, we are testing for optimizer states matching. And if they do not match, then we assert that they are sharded. But the real test would be to compare the aggregated states with the expectation. So, do you think we can do anything to test out the aggregation logic?

Reply:
aggregation on fp32 weights is indirectly tested by asserting that the eval loss matches the expectation, but adding additional checks to compare aggregation logic separately.